### DISCUSSION TITLE
How to avoid conflict with `zi` command from ajeetdsouza/zoxide?

### DISCUSSION THREAD

**Discussion by xfzv** [z-shell/zi#145] [Q&A] [answered]
Out of curiosity, is anyone using both [ajeetdsouza/zoxide](https://github.com/ajeetdsouza/zoxide) and `z-shell/zi`?

`zoxide` also uses `zi` to interactively jump between different directories. As a result, since `zinit` is aliased to `zi`, I cannot figure how to achieve the following:

- Using `zi` only for zoxide interactive jump
- Managing `z-shell/zi` with `zinit` commands only (instead of `zi`)

I know I could use `--cmd <other_letter>` as documented [here](https://github.com/ajeetdsouza/zoxide#flags) but I'm too used to use `z` and `zi` to switch to switch to another letter.

I don't have such issue if using [zdharma-continuum/zinit](https://github.com/zdharma-continuum/zinit) because it doesn't alias `zinit` to `zi` so I can just use for example `zinit self-update` instead of `zi self-update`.

---

- OS: Linux (x86_64)
- Zsh Version: 5.8.1

- `"${ZDOTDIR}/config/plugins/z-shell-zi.zsh"`:
```sh
declare -A ZI
ZI[HOME_DIR]="${ZDOTDIR}/config/plugins/.zi"
ZI[BIN_DIR]="${ZI[HOME_DIR]}/bin"

[[ -f "${ZI[BIN_DIR]}/zi.zsh" ]] && source "${ZI[BIN_DIR]}/zi.zsh" \
    || git clone https://github.com/z-shell/zi "${ZI[BIN_DIR]}"

autoload -Uz _zi
(( ${+_comps} )) && _comps[zi]=_zi

zinit wait lucid for \
 atinit"ZI[COMPINIT_OPTS]=-C; zicompinit; zicdreplay" \
    z-shell/F-Sy-H \
 blockf \
    zsh-users/zsh-completions \
 atload"!_zsh_autosuggest_start" \
    zsh-users/zsh-autosuggestions
```

- `"${ZDOTDIR}/config/plugins/ajeetdsouza-zoxide.zsh"`:
```sh
eval "$(zoxide init zsh)"
```

I also tried to use `eval "$(zoxide init --no-aliases zsh)"` instead and manually defining the following aliases, to no avail:

```sh
alias \
    z="__zoxide_z"  \
    zi="__zoxide_zi"
```

Sourcing zoxide before zi without any alias modification doesn't help much either. In this case, it's the opposite, `zi` works fine for the plugins manager but no longer for `zoxide`.

I'm kind of lost. Thanks in advance for any suggestion! :disguised_face:

**Comment by ss-o** [comment]


**Reply by pskarma** [reply]
I am unaware of zoxide, will check it out.

**Comment by ss-o** [comment]


**Reply by xfzv** [reply]
Can confirm that with https://github.com/z-shell/zi/commit/0df086fd5dc32f4abeffb862bf9e2d1858cebb5b#r71512607, I can use `zplugin` instead of `zinit` while still being able to use `zi` for zoxide interactive jump. :+1: 

What I ended up using based on my few experiments:

- `zoxide` is sourced before `z-shell/zi` and my aliases are sourced last
- Both of the following work to start `zoxide`:

```sh
eval "$(zoxide init zsh)
eval "$(zoxide init --no-aliases zsh)"
```

**as long as I'm using the `zi` alias**:

```sh
alias zi="__zoxide_zi"
```

I guess there might be other methods but I'm fine with this one for now.

**Reply by thomasaarholt** [reply]
Thank you! The collision of zi with zoxide is something that had me really confused!

**Reply by xfzv** [reply]
@thomasaarholt I've just found out that `ZINIT[NO_ALIASES]` option was added a couple of months ago to [zdharma-continuum/zinit](https://github.com/zdharma-continuum/zinit):

- https://github.com/zdharma-continuum/zinit/blob/1375adf8f8edadbdcfc18445d852c7f08e723887/README.md?plain=1#L1046

- [relevant commit](https://github.com/zdharma-continuum/zinit/commit/c624ac00770f7284bc851bea377eedcc974792f4)

If set to `1`, `zi` and `zinit` aliases aren't set. So you could just use:

```console
eval "$(zoxide init zsh)
```

and refrain from explicitly set any `z` / `zi` alias yourself. I've just tried, it works fine.

I'm not sure if `zshell/zi` has this option too, I couldn't find it in the doc.

**Comment by xfzv** [comment]
FYI, after updating to 843cdd386f7820cd16ce5e7ee9c9a002500f9853, I'm now getting `zoxide: no match found` with any `zplugin` command. I guess this is because of these two changes?

- https://github.com/z-shell/zi/commit/843cdd386f7820cd16ce5e7ee9c9a002500f9853#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2678-R2677

    ```diff
    - zplugin() { zi "$@"; }
    + #zplugin() { zi "$@"; }
    ```

- https://github.com/z-shell/zi/commit/843cdd386f7820cd16ce5e7ee9c9a002500f9853#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2701-R2700

    ```diff
    - builtin alias zpl=zi zplg=zi zini=zi zinit=zi
    + builtin alias zpl=zi zplg=zi zini=zi zinit=zi zplugin=zi
    ```

_Edit:_ another workaround is to use `\zi` instead of `zi`, works fine.

**Reply by ss-o** [reply]
My apologies, I am trying to reduce functions that in the past were added as a temporary solution, and it wraps around one another. By aliasing it to zi instead of all zi function expected that it should not affect anyone. 🤦‍♂️

What is the best solution for you? I can reverse it as it was or you might have a better suggestion? 
Both scenarios are fine by me. :octocat:

**Reply by ss-o** [reply]
I had no time to closer look at the zoxide, as technically alias: zplugin=zi and zplugin() { zi "$@"; } are almost same:

### Aliases and Functions ¹

The entire contents of aliases and functions are stored in the shell's memory. A natural consequence of this is aliases and functions can only be used by the current shell, and not by any other programs you may invoke from the shell-like text editors, scripts, or even child instances of the same shell. Aliases and functions are executed by the current shell, i.e. they run within and affect the shell's current environment.² No separate process is necessary to run an alias or function.

### Scripts

Shells do not keep scripts in memory. Instead, scripts are read from the files where they are stored every time they are needed. If the script is found via a $PATH search, many shells store a hash of its pathname in memory to save time on future $PATH look-ups, but that is the extent of a script's memory footprint when not in use.

Scripts can be invoked in more ways than functions and aliases can. They can be passed as an argument to an interpreter, like sh script, or invoked directly as an executable, in which case the interpreter in the shebang line (e.g. #!/bin/sh) is invoked to run it. In both cases, the script is run by a separate interpreter process with its own environment separate from that of your shell, whose environment the script cannot affect in any way. Indeed, the interpreter shell does not even have to match the invoking shell. Because scripts invoked this way appear to behave like any ordinary executable, they can be used by any program.

Finally, a script can be read and run by the current shell with or in some shells, a source. In this case, the script behaves much like a function that is read on-demand instead of being constantly kept in memory.

Application

Given the above, we can come up with some general guidelines for whether to make something a script or function / alias.

- Do other programs besides your shell need to be able to use it? If so, it has to be a script.

- Do you only want it to be available from an interactive shell? It's common to want to change the default behavior of many commands when run interactively without affecting external commands / scripts. For this case, use an alias / function set in the shell's "interactive-mode-only" rc file (for zsh this is .zshrc).

- Does it need to change the shell's environment? Both a function / alias or a sourced script are possible choices.

- Is it something you use frequently? It's probably more efficient to keep it in memory, so make it a function / alias if possible.

Conversely, is it something you use only rarely? In that case, there's no sense having it hog memory when you don't need it, so make it a script.

**Reply by ss-o** [reply]
> FYI, after updating to [843cdd3](https://github.com/z-shell/zi/commit/843cdd386f7820cd16ce5e7ee9c9a002500f9853), I'm now getting `zoxide: no match found` with any `zplugin` command. I guess this is because of these two changes?
> 
>     * [843cdd3#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2678-R2677](https://github.com/z-shell/zi/commit/843cdd386f7820cd16ce5e7ee9c9a002500f9853#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2678-R2677)
>       ```diff
>       - zplugin() { zi "$@"; }
>       + #zplugin() { zi "$@"; }
>       ```
> 
>     * [843cdd3#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2701-R2700](https://github.com/z-shell/zi/commit/843cdd386f7820cd16ce5e7ee9c9a002500f9853#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceL2701-R2700)
>       ```diff
>       - builtin alias zpl=zi zplg=zi zini=zi zinit=zi
>       + builtin alias zpl=zi zplg=zi zini=zi zinit=zi zplugin=zi
>       ```
> 
> 
> _Edit:_ another workaround is to use `\zi` instead of `zi`, works fine.


I don't like workarounds. When all is as it should be - it saves time in the future.

Hmm... I'll take a peek at zoxide. :eyes: :zap:

**Reply by xfzv** [reply]
> My apologies, I am trying to reduce functions that in the past were added as a temporary solution, and it wraps around one another. By aliasing it to zi instead of all zi function expected that it should not affect anyone. man_facepalming

No worries :disguised_face:

> What is the best solution for you? I can reverse it as it was or you might have a better suggestion? Both scenarios are fine by me. :octocat:

See my edit from my previous post, I think that I can simply use `\zi` instead of `zi`. At least it works just fine on my end, without any other modification:

https://user-images.githubusercontent.com/78810647/165852131-ceb8daa2-6fbe-428d-acde-03ba80e22952.mp4

**Comment by ss-o** [comment]


**Reply by xfzv** [reply]
If using this function, I'm getting the following error when opening a shell:

```zsh
/home/xfzv/.config/zsh/config/aliases.zsh:123: defining function based on alias `zi'
/home/xfzv/.config/zsh/config/aliases.zsh:123: parse error near `()'
```

Edit: same error in my `.zshrc`

**Reply by ss-o** [reply]
Ok, what is our target?

command zi: - zoxide jumps
command ??: - ❮ ZI ❯

**Reply by ss-o** [reply]
Also, what installation method do you used for zoxide?

**Reply by ss-o** [reply]
:hear_no_evil: 🤦‍♂️ Now I understand you, I set up the zoxide and its default function is `zi` same as z-shell/zi, and both points "redirects" to the same alias which is also `zi` 🧙‍‍♀️

Which command do you want to be set for z-shell/zi? 🥷

**Reply by xfzv** [reply]
> Ok, what is our target?
> command zi: - zoxide jumps
> command ??: - ❮ ZI ❯

Yes, `zi` for zoxide interactive jumps. I don't have any preference for ❮ ZI ❯ command, I'm fine with anything as long as I can use both together. I can stick with `\zi` too.

> Also, what installation method do you used for zoxide?

I'm using Arch Linux and Gentoo. `zoxide` is installed from their respective package manager (pacman / portage):
- Arch Linux: [community repository](https://archlinux.org/packages/community/x86_64/zoxide/)
- Gentoo: [dm9pZCAq overlay](https://github.com/gentoo-mirror/dm9pZCAq/tree/master/app-shells/zoxide)

Before, I had it installed with `zinit` with [this method](https://github.com/ajeetdsouza/zoxide/issues/175#issuecomment-841470951):

```zsh
zinit ice wait"2" as"command" from"gh-r" lucid \
  mv"zoxide*/zoxide -> zoxide" \
  atclone"./zoxide init zsh > init.zsh" \
  atpull"%atclone" src"init.zsh" nocompile'!'
zinit light ajeetdsouza/zoxide
```

**Comment by ss-o** [comment]


**Reply by ss-o** [reply]
I am thinking about some ASCII-char escape which is 1 or 2 key-press and has native keybinds. 

It's cool, easy, and safe :) And always could be used as a backup. 

Example: alt+31 is ZI symbol - ▼  :octocat: :upside_down_face: and alt+31 seems easy for the hand.

**Comment by ss-o** [comment, marked as answer]
On the branch: [next](https://github.com/z-shell/zi/tree/next) you can alias any command to: ❮▼❯. to be the default.

You pointed out a big issue on how ZI could be compromised with a simple function conflict.

Thank you.

**Reply by xfzv** [reply]
Nice, I've just tried the [new change](https://github.com/z-shell/zi/commit/271357556e8adf466ab3967cc15166e7baf05241#diff-442651f38fa8ad87412388a0b1730a84d180c53d565aa138e906c2c7f93831ceR2677) and it works fine on my end:

```zsh
alias myzialias="❮▼❯"
```

I can simultaneously use `zi` for zoxide interactive jump and `myzialias` for any `z-shell/zi` command.

**Reply by ss-o** [reply]
That's great, this will not disappear this time :see_no_evil: 

I'll merge it soon.

**Reply by iax7** [reply]
the alias didn't work for me. I did this instead:

```bash
eval "$(zoxide init --cmd j zsh)"
```

this will change the `z, zi` commands to `j, ji`

**Reply by ss-o** [reply]
@iax7 your proposal is correct :+1: Additionally I added https://github.com/z-shell/zsh-zoxide. Thanks for sharing :dart:

**Comment by pskarma** [comment]
I also reviewed zoxide as mentioned previously. Great project I am surprised I missed it.
However, I did some comparing as I have my own method of supper fast terminal use.

It's very simple:

- adjust how history is written based on the common sense of my own habits.
- adjust completion and autosuggestions same as above.

And this is basically all. The effect is not only for directories and also for every other input stored in history based on rules set for writing the history, which gets suggested. Based on my activity the tasks performed generally change daily, for this reason, history is refreshed accordingly and specifically excluded for a longer-term. Additionally, I use z-shell/H-S-MW to find everything I pasted or that is long and complex.

The outcome of all this is that it takes 2-4 and rarely more than 4 and almost knows the difference between how complex and long command was.

I chose random tasks and checked, how many key-press it will take, and compared my method and oxide:

zoxide: 324
my method: 112

It may sound funny, but all it takes is just to find the meaning or required settings and almost all can be achieved with native Zsh. I really recommend trying it.

We - humans are very predictable :)

**Reply by timka** [reply]
>    * adjust how history is written based on the common sense of my own habits.
>    * adjust completion and autosuggestions same as above.

Tastes in line with my attitude towards workflow. Could you please elaborate a bit, maybe an example?

**Comment by xfzv** [comment]
Updating Zoxide to 0.9.4 broke `zi` on my end, due to https://github.com/ajeetdsouza/zoxide/commit/55925064567a852d0a96c6e32833836421849c81 which now defines `z` and `zi` as functions instead of aliases:

- https://github.com/ajeetdsouza/zoxide/commit/55925064567a852d0a96c6e32833836421849c81#diff-b1d218e7c366c3393c0700099d104dcfc84757d3150e4fb7731b441c79cbd089R82-R88
- https://github.com/ajeetdsouza/zoxide/commit/55925064567a852d0a96c6e32833836421849c81#diff-b1d218e7c366c3393c0700099d104dcfc84757d3150e4fb7731b441c79cbd089L115-L116

Zoxide is sourced after `zi` and I'm setting `alias zinit="❮▼❯"` last.

Now, any `zinit` command is interpreted as `zi` by Zoxide.

With Zoxide <=0.9.3:

```console
% which zinit
zinit: aliased to ❮▼❯

% alias zinit
zinit=❮▼❯

% which z
z: aliased to __zoxide_z

% which zi
zi: aliased to __zoxide_zi

% zinit list
author/plugin1
author/plugin2
[...]
```

---

With Zoxide 0.9.4:

```console
% which zinit
zinit: aliased to ❮▼❯

% alias zinit
zinit=❮▼❯

% which z
z () {
	__zoxide_z "$@"
}

% which zi
zi () {
	__zoxide_zi "$@"
}

% zinit
# actually runs `zi` i.e. `__zoxide_zi "$@"`

% zinit list
# actually runs `zi list`, i.e. changes directory to the best "list" match in my Zoxide database.
```

Of course, `\zi` won't work either since `zi` is now a function instead of an alias. Basically I cannot perform any `zi` command.

Any idea?
